cli/module_graph.rs

```diff
@@ -0,0 +1,725 @@
+  specifier: String,
+  #[serde(serialize_with = "serialize_module_specifier")]
+  resolved_specifier: ModuleSpecifier,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ModuleGraphFile {
+  pub specifier: String,
+  pub url: String,
+  pub filename: String,
+  pub imports: Vec<ImportDescriptor>,
+  pub referenced_files: Vec<ReferenceDescriptor>,
+  pub lib_directives: Vec<ReferenceDescriptor>,
+  pub types_directives: Vec<ReferenceDescriptor>,
+  pub type_headers: Vec<ReferenceDescriptor>,
+  pub media_type: i32,
+  pub source_code: String,
+}
+
+type SourceFileFuture =
+  Pin<Box<dyn Future<Output = Result<SourceFile, ErrBox>>>>;
+
+pub struct ModuleGraphLoader {
+  permissions: Permissions,
+  file_fetcher: SourceFileFetcher,
+  maybe_import_map: Option<ImportMap>,
+  pending_downloads: FuturesUnordered<SourceFileFuture>,
+  pub graph: ModuleGraph,
+  is_dyn_import: bool,
+  analyze_dynamic_imports: bool,
+}
+
+impl ModuleGraphLoader {
+  pub fn new(
+    file_fetcher: SourceFileFetcher,
+    maybe_import_map: Option<ImportMap>,
+    permissions: Permissions,
+    is_dyn_import: bool,
+    analyze_dynamic_imports: bool,
+  ) -> Self {
+    Self {
+      file_fetcher,
+      permissions,
+      maybe_import_map,
+      pending_downloads: FuturesUnordered::new(),
+      graph: ModuleGraph(HashMap::new()),
+      is_dyn_import,
+      analyze_dynamic_imports,
+    }
+  }
+
+  pub async fn add_to_graph(
+    &mut self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<(), ErrBox> {
+    self.download_module(specifier.clone(), None)?;
+
+    loop {
+      let load_result = self.pending_downloads.next().await.unwrap();
+      let source_file = load_result?;
+      let spec = ModuleSpecifier::from(source_file.url.clone());
+      self.visit_module(&spec, source_file)?;
+      if self.pending_downloads.is_empty() {
+        break;
+      }
+    }
+
+    Ok(())
+  }
+
+  pub fn build_local_graph<S: BuildHasher>(
+    &mut self,
+    _root_name: &str,
+    source_map: &HashMap<String, String, S>,
+  ) -> Result<(), ErrBox> {
+    for (spec, source_code) in source_map.iter() {
+      self.visit_memory_module(spec.to_string(), source_code.to_string())?;
+    }
+
+    Ok(())
+  }
+
+  fn visit_memory_module(
+    &mut self,
+    specifier: String,
+    source_code: String,
+  ) -> Result<(), ErrBox> {
+    let mut imports = vec![];
+    let mut referenced_files = vec![];
+    let mut lib_directives = vec![];
+    let mut types_directives = vec![];
+
+    // let mut dummy_prefix = false;
+
+    // The resolveModules op only handles fully qualified URLs for referrer.
+    // However we will have cases where referrer is "/foo.ts". We add this dummy
+    // prefix "memory://" in order to use resolution logic.
+    let module_specifier =
+      if let Ok(spec) = ModuleSpecifier::resolve_url(&specifier) {
+        spec
+      } else {
+        ModuleSpecifier::resolve_url(&format!("memory://{}", specifier))?
+      };
```

Comment:
This seems very strange.... Why doesn't the normal resolution algorithm handle "/foo.ts" ? Maybe we should just change that.

Reply:
`/foo.ts` cannot be used as a referrer - parsing such URL would return `https://docs.rs/url/2.1.1/url/enum.ParseError.html#variant.RelativeUrlWithCannotBeABaseBase`

Changing this would be a huge undertaking, we need to think this through thoroughly.

Reply:
Hrmmmm... yeah, don't like this easier, but yeah... maybe a `TODO` so we remember to revisit.